src/WebJobs.Extensions.DurableTask/LifeCycleNotificationHelper.cs

```diff
@@ -0,0 +1,243 @@
+        private readonly ExtensionConfigContext extensionConfigContext;
+        private readonly string eventGridKeyValue;
+        private static HttpClient httpClient = null;
+
+        public LifeCycleNotificationHelper(DurableTaskExtension config, ExtensionConfigContext extensionConfigContext)
+        {
+            this.config = config ?? throw new ArgumentNullException(nameof(config));
+            this.extensionConfigContext = extensionConfigContext ?? throw new ArgumentNullException(nameof(extensionConfigContext));
+
+            if (!string.IsNullOrEmpty(config.EventGridTopicEndpoint) && !string.IsNullOrEmpty(config.EventGridKeySettingName))
+            {
+                this.UseTrace = true;
+
+                // Currently, we support Event Grid Custom Topic for notify the lifecycle event of an orchestrator.
+                // For more detail about the Event Grid, please refer this document.
+                // Post to custom topic for Azure Event Grid
+                // https://docs.microsoft.com/en-us/azure/event-grid/post-to-custom-topic
+                httpClient = new HttpClient();
+                INameResolver nameResolver = extensionConfigContext.Config.GetService<INameResolver>();
+                this.eventGridKeyValue = nameResolver.Resolve(config.EventGridKeySettingName);
+                httpClient.DefaultRequestHeaders.Add("aeg-sas-key", this.eventGridKeyValue);
+            }
+        }
+
+        private bool UseTrace { get; }
+
+        public void SetHttpMessageHandler(HttpMessageHandler handler)
+        {
+            httpClient = new HttpClient(handler);
+            httpClient.DefaultRequestHeaders.Add("aeg-sas-key", this.eventGridKeyValue);
+        }
+
+        private async Task SendNotificationAsync(
+            EventGridEvent[] eventGridEventArray,
+            string hubName,
+            string functionName,
+            string version,
+            string instanceId,
+            string reason,
+            FunctionType functionType,
+            bool isReplay,
+            FunctionState functionState)
+        {
+            var json = JsonConvert.SerializeObject(eventGridEventArray);
+            var content = new StringContent(json, Encoding.UTF8, "application/json");
+
+            // Details about the Event Grid REST API
+            // https://docs.microsoft.com/en-us/rest/api/eventgrid/
+            var result = await httpClient.PostAsync(this.config.EventGridTopicEndpoint, content);
+
+            if (!result.IsSuccessStatusCode)
+            {
+                var appName = Environment.GetEnvironmentVariable("WEBSITE_SITE_NAME") ?? string.Empty;
+                var slotName = Environment.GetEnvironmentVariable("WEBSITE_SLOT_NAME") ?? string.Empty;
+                var extensionVersion = FileVersionInfo.GetVersionInfo(typeof(DurableTaskExtension).Assembly.Location).FileVersion;
+
+                this.config.TraceHelper.SendMessageFailed(
```

Comment:
I suggest we always trace, regardless of success or failure. That we customers can know if notifications are actually being sent as part of debugging their Event Grid integration. We should also use a `Stopwatch` to track the latency of these notifications in case there might be a performance impact.

Reply:
Thank you! We will fix it.

Reply:
This seems fixed by @gled4er  :)